### PR TITLE
Feature/3424 archiving project

### PR DESCRIPTION
**Pull request by RK206**
# Upload packages individually

*Please don't delete any sections when completing this PR template; instead enter **N/A** for checkboxes or sections which are not applicable, unless otherwise stated below*

See https://github.com/DOAJ/doajPM/issues/3424 and https://github.com/DOAJ/doajPM/issues/3451

Uploads packages for each journal individually and also load the identifier files irrespective of the case of the file name

## Categorisation

This PR...
- [ ] has scripts to run
- [ ] has migrations to run
- [ ] adds new infrastructure
- [ ] changes the CI pipeline
- [x] affects the public site
- [ ] affects the editorial area
- [x] affects the publisher area
- [ ] affects the monitoring

## Basic PR Checklist

Instructions for developers:
* For each checklist item, if it is N/A to your PR check the N/A box
* For each item that you have done and confirmed for yourself, check Developer box (including if you have checked the N/A box)

Instructions for reviewers:
* For each checklist item that has been confirmed by the Developer, check the Reviewer box if you agree
* For multiple reviewers, feel free to add your own checkbox with your github username next to it if that helps with review tracking

### Code Style

- No deprecated methods are used
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer

- No magic strings/numbers - all strings are in `constants` or `messages` files
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- ES queries are wrapped in a Query object rather than inlined in the code
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer
  
- Where possible our common library functions have been used (e.g. dates manipulated via `dates`)
  - [ ] N/A
  - [ ] Developer
  - [ ] Reviewer
  
- Cleaned up commented out code, etc
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer

### Testing

- Unit tests have been added/modified
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- Functional tests have been added/modified
  - [ ] N/A
  - [x] Developer
  - [x] Reviewer
  
- Code has been run manually in development, and functional tests followed locally
  - [ ] N/A
  - [x] Developer
  - [ ] Reviewer

### Documentation

- FeatureMap annotations have been added
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer
  
- Documentation updates - if needed - have been identified and prepared for inclusion into main documentation (e.g. added and highlighted/commented as appropriate to this PR)
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer
  
- Core model documentation has been added to if needed: https://docs.google.com/spreadsheets/d/1lun2S9vwGbyfy3WjIjgXBm05D-3wWDZ4bp8xiIYfImM/edit
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer

- Events and consumers documentation has been added if needed: https://docs.google.com/spreadsheets/d/1oIeG5vg-blm2MZCE-7YhwulUlSz6TOUeY8jAftdP9JE/edit
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer
  
- The docs for this branch have been generated and pushed to the doc site (see docs/README.md for details)
  - [ ] N/A
  - [ ] Developer
  - [x] Reviewer

### Release Readiness

- If needed, migration has been created and tested locally
  - [x] N/A
  - [ ] Developer
  - [x] Reviewer

- Release sheet has been created, and completed as far as is possible https://docs.google.com/spreadsheets/d/1Bqx23J1MwXzjrmAygbqlU3YHxN1Wf7zkkRv14eTVLZQ/edit
  - [ ] N/A
  - [ ] Developer
  - [ ] Reviewer

- There has been a recent merge up from `develop` (or other base branch).  List the dates of the merges up from develop below
  - [date of merge up]

## Testing

List the Functional Tests that must be run to confirm this feature

1. [Upload correctly structured file with multiple journals and it is processed correctly](https://doaj.github.io/doaj-docs/feature/3424_archiving_project/testbook/index.html#articles_preservation/user_test_script/upload_correctly_structured_file_with_multiple_journals_and_it_is_processed_correctly)
2. Preservation details need to be added in the config file on the test server

## Deployment

What deployment considerations are there? (delete any sections you don't need)

### Configuration changes

What configuration changes are included in this PR, and do we need to set specific values for production

### Scripts

What scripts need to be run from the PR (e.g. if this is a report generating feature), and when (once, regularly, etc).

### Migrations

What migrations need to be run to deploy this

### Monitoring

What additional monitoring is required of the application as a result of this feature

### New Infrastructure

What new infrastructure does this PR require (e.g. new services that need to run on the back-end).

### Continuous Integration

What CI changes are required for this

### CODE REVIEW

**Review by richard-jones — CHANGES_REQUESTED** [review]
Can we get an update to one or more of the functional tests to show off the new behaviour?  Can that include one or more test packages to go along with the test.

For example, add another one of these for an upload with multiple journals: https://doaj.github.io/doaj-docs/develop/testbook/index.html#articles_preservation/user_test_script/upload_correctly__structured_file_and_it_is_processed_correctly

**Comment by RK206** [comment]
I have updated the functional testcase for multiple journals

File: doajtest/testbook/articles_preservation/user_test_script.yml [hunk 1]
```diff
@@ -75,3 +75,27 @@ tests:
...
       successful will be displayed
+- title: Upload correctly structured file with multiple journals and it is processed correctly
+  context:
+    role: publisher
+  steps:
+  - step: Ensure that the archive package has been created with correct structure
+  - step: Ensure that the archive package contains more than one article with different journals. 
+          A sample zip file is available for testing purpose. If the articles provided in zip are not available in the elasticsearch data, 
+          the results may not be as expected. In that case unzip the file and update the identifiers.txt file with correct identifier for the articles.
+    resource: /preservation_upload_test_package/multi_journals.zip
```
Comment (richard-jones) [inline doajtest/testbook/articles_preservation/user_test_script.yml:86]: Can we also add an example package resource to the other relevant tests:

https://doaj.github.io/doaj-docs/feature/3424_archiving_project/testbook/index.html#articles_preservation/user_test_script/upload_correctly__structured_file_and_it_is_processed_correctly

https://doaj.github.io/doaj-docs/feature/3424_archiving_project/testbook/index.html#articles_preservation/user_test_script/upload_incorrectly__structured_file

https://doaj.github.io/doaj-docs/feature/3424_archiving_project/testbook/index.html#articles_preservation/user_test_script/upload_correctly__structured_file_with_articles_that_user_do_not_own

**Review by richard-jones — CHANGES_REQUESTED** [review]
@RK206 I didn't have time to complete the review yet, I will do that asap, but I ran out of time today but there were a couple of changes to request, so I'm submitting this review as is for the moment.

**Comment by RK206** [comment]
Implemented the changes as suggested

File: doajtest/unit/test_task_preservation.py [hunk 2]
```diff
@@ -23,14 +23,14 @@ def setUp(self):
 
         self.upload_dir = app.config.get("UPLOAD_DIR", ".")
         created_time = dates.format(datetime.utcnow(), "%Y-%m-%d-%H-%M-%S")
-        owner = "rama"
-        dir_name = owner + "-" + created_time
+        self.owner = "rama"
```
[the remark this answers is not in the document]
Comment (richard-jones) [reply doajtest/unit/test_task_preservation.py:62]: I found the multi-journals test, and I found a couple of bugs in it, so I'm not sure if you were running this locally.  I needed to change "identifier" to "Identifier" in a couple of places, then the test passed.  Looking at the test, though, I still don't know how to confirm if it's testing the multiple packages thing, can you walk me through it?

File: doajtest/testbook/articles_preservation/upload_preservation_files.yml [hunk 3]
```diff
@@ -75,11 +79,40 @@ tests:
...
+      to be processed.'
+    - Upload status is shown in 'History of uploads' with status processing
+  - step: wait a short amount of time for the job to process, then reload the page
+      (do not re-submit the form data).  If the job remains in "pending", reload the
+      page until the status changes.
+    results:
+    - Upload status in 'History of uploads' changes to 'successfully processed' and
+      the 'Notes' shows as 'uploaded to archive server'
+    - The 'Notes' has the label 'UPLOADED PACKAGES'
+    - Under the label 'UPLOADED PACKAGES', details of the packages that were uploaded are displayed in the format <journal_id>_<Date>.tar.gz
```
Comment (richard-jones) [inline doajtest/testbook/articles_preservation/upload_preservation_files.yml:110]: I don't see this, instead I get:

```
Uploaded articles to archive server

Successful articles

article1, article3, article2
```

**Review by richard-jones — CHANGES_REQUESTED** [review]
There's one common library thing to change (use of `dates.now()`), and a comment about the multi-package unit test.

Otherwise, I am able to run the functional test locally, and it succeeds, but I don't know how to tell whether the uploaded packages are correct in IA, can you add some testing instructions to the PR and/or the functional test to cover how to check what was uploaded to IA?

**Review by richard-jones — APPROVED** [review]
I was able to to successfully test this, and I have pushed a small text update to the functional test to aid the DOAJ team setting up for testing

**Pull request closed without merging**